Make DataBox non-copyable

wthrowe opened the pull request:
Copying a DataBox just created aliases between all the members, which resulted in the surprising situation that mutating one DataBox would also change the other.  As a side-effect, this allowed const DataBoxes to be mutated by creating a non-const aliasing version.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Make DataBox non-copyable

Copying a DataBox just created aliases between all the members, which
resulted in the surprising situation that mutating one DataBox would
also change the other.  As a side-effect, this allowed const DataBoxes
to be mutated by creating a non-const aliasing version.

The only uses of DataBox copying were in three tests, two of which
were bugs and the third of which was verifying that copying a DataBox
did not break something and so is now irrelevant.

nilsdeppe commented:
This behavior is intentional because I'm fairly certain it's what we want for observers. This way you can add observers to the DataBox without recomputing or copying everything else over. I don't think this should get merged until we have a clear path forward there.

wthrowe committed:
Disallow mutable aliasing from db::create_from

Allowing a DataBox produced via create_from to be mutated could
unexpectedly change the source DataBox, as described for the copy
constructor in the preceding commit.

This change forces all aliasing copies to be const temporaries (whose
lifetimes can be extended in the usual way), so their contents cannot
be mutated to change the state of the initial DataBox.

There were no non-trivial uses of the removed functionality.

wthrowe commented:
Right, I'd forgotten about that use case.  How about this new version, where you can make aliasing boxes (and add/remove things from them) using db::create_from, but the resulting boxes are always const so their contents cannot be mutated?

kidder commented:
@nilsdeppe please look at

wthrowe commented:
(Hopefully) fixed Travis.  Added a test that create_from returns the correct type.
